### PR TITLE
DBZ-720 snapshotting

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://issues.jboss.org/browse/DBZ-720

@jpechane And here's the snapshotting implementation for Oracle. Based on top of #8, only the last commit is new.

### CODE REVIEW

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java [hunk 1]
```diff
@@ -81,6 +82,16 @@
             .withValidation(Field::isRequired)
             .withDescription("Name of the XStream Out server to connect to.");
 
+    public static final Field SNAPSHOT_MODE = Field.create("snapshot.mode")
```
Comment (jpechane) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java]: This should belong to the parent class

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/SnapshotChangeRecordEmitter.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import io.debezium.relational.RelationalChangeRecordEmitter;
+import io.debezium.util.Clock;
+import oracle.streams.RowLCR;
+
+/**
+ * Emits change data based on a single {@link RowLCR} event.
+ *
+ * @author Gunnar Morling
+ */
+public class SnapshotChangeRecordEmitter extends RelationalChangeRecordEmitter {
```
Comment (jpechane) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/SnapshotChangeRecordEmitter.java]: Is this class specific to oracle?

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java]: I was wondering that, too, but then I figured that the actual snapshotting modes are different between the different connectors. Hence the idea of `SnapshottingTask`, based on its specific snapshotting modes and the previous offset (if any), a specific sub-class returns this descriptor which indicates whether a snapshot of schema, data or both is needed.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/SnapshotChangeRecordEmitter.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/SnapshotChangeRecordEmitter.java]: I suppose that it's not. Let's move it to core once it's proven with SQL Server.

**Comment by gunnarmorling** [comment]
@jpechane Rebased to current master.

**Comment by jpechane** [comment]
@gunnarmorling Could yoy please check `OracleDatabaseSchema`? It seems that an unused imports crept in in a previous commits which fails build.
Also `StreamingDatatypesIT` and `SnapshotDatatypesIT` does not work as records created as part of `io.debezium.connector.oracle.OracleConnectorIT.shouldTakeSnapshot()` seems to stay in database and are delivered to the aforementioned tests.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleOffsetContext.java [hunk 3]
```diff
@@ -64,4 +123,62 @@ public void setTransactionId(String transactionId) {
...
+        }
+
+        @Override
+        public Map<String, ?> getPartition() {
+            return Collections.singletonMap(SERVER_PARTITION_KEY, logicalName);
+        }
+
+        @Override
+        public OffsetContext load(Map<String, ?> offset) {
+            long scn = (long) offset.get(SourceInfo.SCN_KEY);
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleOffsetContext.java]: Interesting, I'm not geting this failure. But I'll change it regardless.

[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleOffsetContext.java]: The one I get is
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project debezium-connector-oracle: Compilation failure
[ERROR] /home/jpechane/sources/debezium/debezium-incubator/debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleOffsetContext.java:[177,41] incompatible types: capture#1 of ? cannot be converted to long
```

**Comment by gunnarmorling** [comment]
Out of curiousity, which JDK version are you on?

**Comment by jpechane** [comment]
```
openjdk version "1.8.0_171"
OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-0ubuntu0.16.04.1-b11)
OpenJDK 64-Bit Server VM (build 25.171-b11, mixed mode)
```

**Comment by gunnarmorling** [comment]
@jpechane Ok, I've pushed one more commit which I hope fixes all the issues. Build times (well, testing times) are killing me, though :(

**Comment by jpechane** [comment]
@gunnarmorling Thanks a lot it works now, applied!

**Pull request merged**